dev/core#5838 Formbuilder autofill case

jaapjansma opened the pull request:
Overview
----------------------------------------

It is not possible to autofil a Case on a form builder based on the case being viewed.

There are two use cases when this could be useful:
- When the form is shown on the manage case screen
- When a person (anonymous user) receives a link to the form by email which is send from the manage case screen (this is possible with PR #32599)

Before
----------------------------------------

Not possible to load a case based the case being viewed

After
----------------------------------------

Autofil the case. 

![2025-04-11_17-44](https://github.com/user-attachments/assets/b52a343c-e9f3-4b08-9b05-7fb139f93918)

colemanw commented on ext/civi_case/Civi/Afform/Behavior/CaseAutofill.php:
```diff
@@ -0,0 +1,71 @@
+    return E::ts('Autofill');
+  }
+
+  public static function getKey():string {
+    // Would be contact-autofill but this supports legacy afforms from before this was converted to a behavior
+    return 'case-autofill';
+  }
```
This override is not necessary.

colemanw commented:
The error crashing the tests is
```
Error: Class "Civi\Afform\AbstractBehavior" not found in include() (line 13 of /home/homer/buildkit/build/build-4/web/sites/all/modules/civicrm/ext/civi_case/Civi/Afform/Behavior/CaseAutofill.php).
```

What's happening is that the new `CaseAutofill` class extends a class which is provided by the `afform` extension.

Random guess: perhaps since `civi_case` is an extension tied to a component, it gets loaded before the `afform` extension and for a brief instant the class is not available?

@totten any ideas?

jaapjansma commented:
@colemanw and @totten I have updated the PR and moved away from the `Autoservice` and implemented `CompilerPass` class instead with a condition in it whether the Afform extension is installed. If the afform extension is installed the behavior is added to the container. If not no class is loaded.

totten commented:
That's pretty tricky.

So just observing/interpreting: the idea is for `civi_case` to have a _soft/undeclared dependency_ on `afform`. Specifically,  `ext/civi_case` supplies `CaseAutofill.php` which depends on `ext/afform`'s `AfformBehavior.php` --  but `civi_case` does __not__ officially `<require>` the `afform` extension.

But... what if we were thinking about the dependency-graph more explicitly... which way the arrows point? We also have soft-dependencies in the other direction (i.e. `ext/afform/core/managed/AfformPlacement.mgd.php` and `ext/afform/admin/afformEntities/` have preexisting dependencies on CiviCase). If we create a dependency from `civi_case` to `afform`, then the dependency-graph becomes cyclical. (Modules are usually easier to understand when their dependencies are acyclic, aka [DAGs](https://en.wikipedia.org/wiki/Directed_acyclic_graph).)

Just to ask the dumb question - would it be OK to move `CaseAutofill` from `ext/civi_case/` to `ext/afform/core`?  In that way, we get a cleaner DAG (*the new arrows point the same way as the existing arrows*), and we make it impossible for anyone to try to load `CaseAutofill`  without access to `AfformBehavior`?

totten commented:
> Just to ask the dumb question - would it be OK to move `CaseAutofill` from `ext/civi_case/` to `ext/afform/core`?

If the answer is "*Yes*", then that probably makes all the other issues go away? That feels like the most KISS contingency.

If the answer is "*No, it must stay in `ext/civi_case`*"... then right, we are looking at conditionality+guards in the scanning process. Some free-association/thinking-out-loud:

* Ah, interesting, I wouldn't have thought of `CompilerPass` as way to skip. However, I guess the problem is that it only tells the `Container` to skip. The `ClassScanner` still tries to evaluate the class (and hence the additional commit to hide errors in `ClassScanner`).

* I'm a little nervous about hiding the errors in `ClassScanner`. Wouldn't that be expansive about suppressing code-flaws? Ex: You create `MyClass.php` and declare `MyClass extends SomeBaseClasz` with a typo in the base-class. It suppresses the error and acts as if `MyClass` doesn't exist at all. In testing, you find that `MyClass` is simply never used -- and there's no indication for why. Right?

* It would be quite pretty if `CaseAutofill.php` started with a guard like:

    ```php
    if (!class_exists(AfformBehavior::class)) { return; }`
    class CaseAutofill extends AfformBehavior { ... }
    ```

    However, I have doubts. When confronted with various flows of toggling the extension (*esp for testing*), I think the behavior would feel erratic. (*The classloader tends to use `require_once`, and so the behavior be quite sensitive to the start-state when page-load begins.*)

* Another way to achieve conditionality:

    In `info.xml`, disable `scan-classes@1` and replace it with an equivalent implementation of `hook_civicrm_scanClasses`, e.g.

    ```php
    function civi_case_civicrm_scanClasses(array &$classes) {
      ClassScanner::scanFolders($classes, __DIR__ . '/', 'Civi', '\\');
    }
    ```

    To include or exclude some files, you can revise that statement, e.g.

    ```php
    function civi_case_civicrm_scanClasses(array &$classes) {
      $extMap = CRM_Extension_System::singleton()->getMapper();
      $excludeRegex = $extMap->isActiveModule('afform') ? NULL : '/CaseAutofill/'
      ClassScanner::scanFolders($classes, __DIR__ . '/', 'Civi', '\\', $excludeRegex);
    }
    ```

    That should allow it to use `AutoService` / `AfformBehavior` without needing any patches to `ClassScanner` or any direct references to `Container` APIs.

* There may be a better protocol for how to choose classes to scan (e.g. some revision to `scan-classes@1`  or some replacement like a `scan-classes@2`). I'm not sure what that is. My gut says to wait for some more concrete problems before trying to change it generically.

jaapjansma commented:
@totten thanks. I don't think it is a good idea to move this code to afform. I have two reasons for that. One is that we then get a conditional in afform whether CiviCase is enabled. This also came up in PR #32594 where we decided to move similar functionality to the CiviCase extension.

I do love your suggestion of using a custom implementation of `hook_civicrm_scanClasses` rather than the mixin. Your suggestion with excluding is also quite elegant and small.

It would be nice to have some sort of generic approach for this but that is out of scope of this PR.

seamuslee001 commented:
I agree that I think this should be in the CiviCase Extension, I agree with Tim that we shouldn't be trying to hide class errors but I can also see Jaap's point of having a generic approach because I'm sure that this won't be the last time we see these sorts of issues. I would imagine that if someone tried disabling civi_case that form builder / search kit would need to be able to gracefully handle that if the class is no longer accessible as well right?

jaapjansma commented:
I have updated the PR

jaapjansma commented:
I have also created a PR #32602 with a generic proposal.

aydun commented:
>  civi_case does not officially <require> the afform extension.

At this point in the evolution of afform and core, is there are any benefit to not making afform required by core?

totten commented:
> (@jaapjansma) I have updated the PR 

Cool. Personally, assuming the rest of the functionality is working, then I think the current technique is mergeable. (And it can replaced/revised by other PRs.)

> > (@aydun) At this point in the evolution of afform and core, is there are any benefit to not making afform required by core?
>
> (@colemanw) to clarify, Afform is [required by core](https://github.com/civicrm/civicrm-core/blob/066b0909f41e6bb33156dc669e2bc943971cfff6/ext/afform/core/info.xml#L22). But I think this bug still comes up due to the order in which extensions load.

I read this PR with a later revision/test-result than you, so I'm not certain the best way to see the problem from that angle. But the general idea sounds plausible.. e.g.. "Afform has moved into the core-requirements realm but isn't actually loaded at some moment during testing".

I don't know how there would be a difference in runtime-ordering, but this reminds me of an important distinction around installation process. Compare these files/folders:

* `sql/civicrm_data/`: This defines the _universal baseline data_ (*affecting all installations and all tests*). This is essentially a static declaration with no inputs. It includes some default entries for [`civicrm_extension.sqldata.php`](https://github.com/civicrm/civicrm-core/blob/6.2/sql/civicrm_data/civicrm_extension.sqldata.php). It has no way to fire `hook_install`; it only works properly with stateless extensions.
* `setup/`: This defines installation logic. It is used for new/real sites and some tests (_but not all tests_). This is essentially dynamic (accepting parameters like `cmsBaseUrl` and `templateCompilePath`). It does many things, including...
    * Setting up the file-tree
    * Loading `sql/civicrm_data/`
    * [Enabling default extensions](https://github.com/civicrm/civicrm-core/blob/6.2/setup/plugins/init/DefaultExtensions.civi-setup.php), which includes firing `hook_install`, and which allows a mix of stateless/stateful.
 
There are notes about the similarities/differences in the header of [`civicrm_extension.sqldata.php`](https://github.com/civicrm/civicrm-core/blob/6.2/sql/civicrm_data/civicrm_extension.sqldata.php).

For `civi_case` and `afform`, it's probably relevant to note:

* __`civi_case` is stateless.__ It originated as a `component`. [It is initially activated through `sql/civicrm_data/`](https://github.com/civicrm/civicrm-core/blob/6.2/sql/civicrm_data/civicrm_extension.sqldata.php#L61)).
* __`afform` is stateful.__ It has a SQL table and upgrade-steps. It originated as an extension. [It is initially activated through `setup/`.](https://github.com/civicrm/civicrm-core/blob/6.2/setup/plugins/init/DefaultExtensions.civi-setup.php#L21)

jaapjansma commented:
@colemanw and @totten can you help me with moving this PR forward?

jaapjansma committed:
dev/core#5838 formbuilder autofil case

briangflett commented:
## CiviCRM 6.5.0 Compatibility Issue

  Despite being merged on July 9, 2025, this PR did not make it into CiviCRM 6.5.0 (released August
   7, 2025).

  **Issue:** The patch fails to apply on CiviCRM 6.5.0 due to a line number conflict in
  `ext/civi_case/info.xml`.

  **Root Cause:** CiviCRM 6.5.0 has an additional `<mixin>setting-php@1.0.0</mixin>` entry that
  shifts line numbers.

  **Fix Required:** Update the patch's `info.xml` hunk from:
  ```diff
  @@ -29,7 +29,6 @@
     </classloader>
     <mixins>
       <mixin>ang-php@1.0.0</mixin>
  -    <mixin>scan-classes@1.0.0</mixin>
     </mixins>

  To:
  @@ -31,7 +31,6 @@
       <mixin>ang-php@1.0.0</mixin>
  -    <mixin>scan-classes@1.0.0</mixin>
       <mixin>setting-php@1.0.0</mixin>
     </mixins>

  This prevents users from having to manually resolve the conflict when applying these changes to
  6.5.0 installations.
